File: datacapturing/src/main/java/de/cyface/datacapturing/backend/GeoLocationStatusListener.java
package de.cyface.datacapturing.backend;

import android.location.GpsStatus;
import android.location.LocationManager;

/**
 * Implementation of a <code>GeoLocationDeviceStatusHandler</code> used for devices with Android prior to Nougat (API
 * 24).
 *
 * @author Klemens Muthmann
 * @version 2.0.2
 * @since 1.0.0
 * @see GnssStatusCallback
 */
public class GeoLocationStatusListener extends GeoLocationDeviceStatusHandler {
    /**
     * The Android system listener wrapped by this class.
     */
    private final GpsStatus.Listener listener;

    /**
     * Creates a new completely initialized <code>GeoLocationStatusListener</code>.
     *
     * @param manager The <code>LocationManager</code> used to get geo location status updates.
     * @throws SecurityException If fine location permission has not been granted.
     */
    GeoLocationStatusListener(final LocationManager manager) throws SecurityException {
        super(manager);
        this.listener = new GpsStatus.Listener() {
            @Override
            public void onGpsStatusChanged(int event) {
                switch (event) {
                    case GpsStatus.GPS_EVENT_SATELLITE_STATUS:
                        handleSatelliteStatusChange();
                        break;
                    case GpsStatus.GPS_EVENT_FIRST_FIX:
                        handleFirstFix();
                        break;
                }
            }
        };
        locationManager.addGpsStatusListener(this.listener);
    }

    @Override
    public void shutdown() {
        locationManager.removeGpsStatusListener(listener);
    }
}

File: datacapturing/src/main/java/de/cyface/datacapturing/backend/GeoLocationStatusListener.kt
package de.cyface.datacapturing.backend

import android.annotation.SuppressLint
import android.location.GpsStatus
import android.location.LocationManager

/**
 * Implementation of a `GeoLocationDeviceStatusHandler` used for devices with Android prior to Nougat (API
 * 24).
 *
 * @author Klemens Muthmann
 * @version 2.0.2
 * @since 1.0.0
 * @see GnssStatusCallback
 * @param manager The `LocationManager` used to get geo location status updates.
 * @throws SecurityException If fine location permission has not been granted.
 */
@SuppressLint("MissingPermission") // UI has to ensure permissions are available
class GeoLocationStatusListener internal constructor(manager: LocationManager?) :
    GeoLocationDeviceStatusHandler(manager!!) {
    /**
     * The Android system listener wrapped by this class.
     */
    private val listener = GpsStatus.Listener { event ->
        when (event) {
            GpsStatus.GPS_EVENT_SATELLITE_STATUS -> handleSatelliteStatusChange()
            GpsStatus.GPS_EVENT_FIRST_FIX -> handleFirstFix()
        }
    }

    init {
        locationManager.addGpsStatusListener(this.listener)
    }

    override fun shutdown() {
        locationManager.removeGpsStatusListener(listener)
    }
}
